### PR TITLE
implementation of uni-card help feature

### PR DESCRIPTION
**Pull request by PatrykSzwer**
Needed for https://github.com/Starcounter/Blending/issues/163

This is my proposition of implementation additional help feature for `uni-card`.

Working example can be seen on https://github.com/Starcounter/Blending/tree/163_inline-help Blending app branch in `CompositionEditor` page.

How it looks:
repro

How it is implemented in Blending:
```
<div class="uni-card">
                 <slot name="blendingeditor/compositioneditor-description-paragraph"></slot>
                 <input slot="help" id="help" type="checkbox" />
                 <label for="help">
                     Help
                 </label>
                 <div slot="help-message">
                     For more information about using the composition editor, see the docs: <a href="https://docs.starcounter.io/guides/web-apps/view-composing" target="_blank">https://docs.starcounter.io/guides/web-apps/view-composing</a>.
                 </div>
             </div>
```

I don't think that this implementation is the cleanest that It could be but TBH I don't have any good ideas how to make is simpler enough. If you have some suggestions about that issue, @alshakero, @tomalec, @warpech, I would appreciate them.

After we agree and merge implementation of this feature, I'll add it to all or most of the `uni-card` in Blending app.

### CODE REVIEW

**Comment by alshakero** [comment]
Nice work, I know the checkbox suggestion was mine. But don't you think this one might be cleaner?

https://jsbin.com/kamiboniko/1/edit?html,css,output

**Comment by alshakero** [comment]
I know the checkbox idea was mine, but I think this one might be cleaner and more accessible https://codepen.io/alshakero/pen/zjxPQQ

**Comment by PatrykSzwer** [comment]
Looks cool, I'll try this out!

**Comment by PatrykSzwer** [comment]
I've changed implementation to the one using `details`. Example in Blending app was also changed.

**Comment by alshakero** [comment]
It's working perfect but doesn't look good. I'll take a spin at it.

**Comment by alshakero** [comment]
I think it's much better now 

connecting

**Comment by PatrykSzwer** [comment]
I've done small fix and I think It's final version, @warpech let us know what you think.
Demo:
rep

File: uniform.css [hunk 1]
```diff
@@ -445,6 +445,48 @@ Pagination
   padding: 16px;
   box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.4);
 }
+
+.uni-card > details summary {
+  text-align: right;
```
Comment (warpech) [inline uniform.css]: Please add `cursor: pointer` to give indication that this is clickable.

**Comment by alshakero** [comment]
> Does the classless API of details & summary implemented here not limit us from more sophisticated usages as presented in Material Design?

I think the biggest limitation we have right now is the disability to add more than one button to expand. We're locked into one button. I tried many ways, and I think it's not possible to have two `details` inlined. The only options is stacked and it didn't look nice, especially when you expand the top one, it pushes the second button down.

File: uniform.css [hunk 2]
```diff
@@ -445,6 +445,48 @@ Pagination
...
+  box-sizing: border-box;
+  outline: none;
```
Comment (tomalec) [inline uniform.css]: This may break the accessibility of this element https://a11yproject.com/posts/never-remove-css-outlines/

File: uniform.css [hunk 3]
```diff
@@ -445,6 +445,48 @@ Pagination
...
+  padding: 6px;
+
+  -webkit-touch-callout: none;
+  -webkit-user-select: none;
+  -khtml-user-select: none;
+  -moz-user-select: none;
+  -ms-user-select: none;
+  user-select: none;
```
Comment (tomalec) [inline uniform.css]: Why is it nonselectable?

**Comment by tomalec** [comment]
```html
<div class="uni-card">
                 <slot name="blendingeditor/compositioneditor-description-paragraph"></slot>
                 <input slot="help" id="help" type="checkbox" />
                 <label for="help">
                     Help
                 </label>
                 <div slot="help-message">
                     For more information about using the composition editor, see the docs: <a href="https://docs.starcounter.io/guides/web-apps/view-composing" target="_blank">https://docs.starcounter.io/guides/web-apps/view-composing</a>.
                 </div>
             </div>
```
This is not intended use, is it? Help message content and label should come from the outter tree as well.

**Comment by PatrykSzwer** [comment]
> This is not intended use, is it? Help message content and label should come from the outter tree as well.
No It's not. Current usage looks like this:
```
                <details>
                    <summary>Help</summary>
                    <div>
                        Help text
                    </div>
                </details>
```

File: uniform.css (hunk 3 above)
Comment (PatrykSzwer) [reply uniform.css]: To remove annoying highlighting text when you're clicking on summary text to expand help message.

Comment (warpech) [reply uniform.css]: I tried the example at https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details and I don't see any annoying highlighting text.

I agree with @tomalec that we should not compromise web page accessibility by making things outline-less or non-selectable.

**Comment by warpech** [comment]
> I think the biggest limitation we have right now is the disability to add more than one button to expand. We're locked into one button. I tried many ways, and I think it's not possible to have two details inlined. The only options is stacked and it didn't look nice, especially when you expand the top one, it pushes the second button down, below the top one's expanded part.

I think that's not a problem.

If there are two different kinds of details, they can be rendered in the same `<details>` tag using a `dom-if`.

File: uniform.css (hunk 3 above)
Comment (alshakero) [reply uniform.css]: Did you try on Chrome?
selection

**Comment by warpech** [comment]
As I said to @PatrykSzwer this morning, I think that the top and bottom padding of the expanded `<detail>` is too small narrow.

Compare ours:

card help 1

With material design components from the links in https://github.com/Starcounter/uniform.css/pull/77#pullrequestreview-114216318:

image

card help 2

**Comment by PatrykSzwer** [comment]
>As I said to @PatrykSzwer this morning, I think that the top and bottom padding of the expanded <detail> is too small narrow.

I spoke with @alshakero about this and he agreed. He will add that additional space.

**Comment by alshakero** [comment]
Good?
selection2

**Comment by PatrykSzwer** [comment]
We decided that the best look and feel comes with no paddings in element and <p> tags used in text. Please remove those new paddings, sorry for that.

**Comment by PatrykSzwer** [comment]
I still think we should find some way to remove this blue outline, It looks just dramatically ugly.

**Comment by alshakero** [comment]
I undid my last commit. Everything is back as it was now.

**Comment by PatrykSzwer** [comment]
@tomalec do you have any idea how could we get rid of this blue outline in some safe way?

**Comment by tomalec** [comment]
> @tomalec do you have any idea how could we get rid of this blue outline in some safe way?

Safest and simples way: don't remove it :)

Seriously, I think we should keep it. We could make it more subtle.
https://developer.mozilla.org/en-US/docs/Web/CSS/outline
Or if you are brave enough use `:focus-ring`
https://www.youtube.com/watch?v=ilj2P5-5CjI

**Comment by alshakero** [comment]
I think this is the best compromise: selection2

**Comment by PatrykSzwer** [comment]
Good enough for me. (can't give approve for my own PR)

**Comment by tomalec** [comment]
Why you removed travis configuration? I wouldn't say travis CI is useless https://travis-ci.org/Starcounter/uniform.css

**Comment by alshakero** [comment]
My bad. I forgot about components and thought it's old travis file from the days where there was a build step. I'll PR it back in a min

**Pull request merged**